#### Student Performance Indicator

Life cycle of Machine Learning project

* Understanding the Problem Statement
* Data Colleciton
* Data Checks to Perform
* Exploratory Data Analaysis (EDA)
* Data Pre-Processing
* Model Training
* Choose Best Model

#### 1) Problem Statement

* This project understands how the student's performance (test scores) is affected by other variables such as gender, ethnicity, parental level of education, lunch, and test preparation course.

#### 2) Data Collection

* Dataset source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
* Data consists of 8 columns and 1000 rows.

#### 2.1 Importing Data and Required Packages
###### Importing Pandas, Numpy, Matplotlib, Seaborn, and Warings Library

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

##### Importing CSV Data as Pandas Dataframe

In [39]:
df = pd.read_csv('data/stud.csv')

Previewing the Data

In [4]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

Viewing the shape of the dataset (rows, columns)

In [6]:
df.shape

(1000, 8)

##### __2.2 Dataset Information__

* gender: sex of students -> [male, female]
* race/ethnicity: ethnicity of students -> [Group A/B/C/D/E]
* parental level of education: parent's highest level of education -> [bachelor's degree, some college, master's degree, associate's degree, high school]
* lunch: type of lunch prior to testing -> [standard, free/reduced]
* test preparation course: completed or not prior to testing -> [completed, none]
* math score -> [0:100]
* reading score -> [0:100]
* writing score -> [0:100]

##### __3. Data Checks to Perform__

* Check for missing values
* Check for duplicates
* Check data types
* Check frequency of unique values for each feature
* Check general statistics of the dataset
* Check various categories present in the different categorical columns

##### __3.1 Check Missing Values__

In [7]:
df.isna().sum()
# isna = is not available

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

Therefore, we can see that there are no missing values in this data set.

##### __3.2 Check for Duplicates__

In [8]:
df.duplicated().sum()

0

Therfore, there are no duplicates values in this dataset.

##### __3.3 Check Data Types__

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


##### __3.4 Check the number of unique values for each column__

In [11]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

##### __3.5 Check the general statistics of the data set__

In [12]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

__Insight__

* from the above description of numerical data, all the means are very similar, ranging between 66 and 69.17
* all standard deviations are also very similar, between 14.6 and 15.2
* math has the lowest minimum score at 0, while writing is higher at 10, with reading the highest at 17

##### __3.7 Exploring Data__

In [13]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [20]:
df.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [27]:
# Listing the categories for each feature:

for col in df.columns:
    print(f"Categories in '{col}' variable: ", end="")
    print(df[f'{col}'].unique())
    # print('\n')

# print("Categories in 'gender' variable: ", end = "")
# print(df['gender'].unique())

Categories in 'gender' variable: ['female' 'male']
Categories in 'race/ethnicity' variable: ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental level of education' variable: ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable: ['standard' 'free/reduced']
Categories in 'test preparation course' variable: ['none' 'completed']
Categories in 'math score' variable: [ 72  69  90  47  76  71  88  40  64  38  58  65  78  50  18  46  54  66
  44  74  73  67  70  62  63  56  97  81  75  57  55  53  59  82  77  33
  52   0  79  39  45  60  61  41  49  30  80  42  27  43  68  85  98  87
  51  99  84  91  83  89  22 100  96  94  48  35  34  86  92  37  28  24
  26  95  36  29  32  93  19  23   8]
Categories in 'reading score' variable: [ 72  90  95  57  78  83  43  64  60  54  52  81  53  75  89  32  42  58
  69  73  71  74  70  65  87  56  61  84  55  44  41  85  59  17  39  80
  37  63

In [33]:
# Define numerical and categorical features
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'object']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'object']

print(f"We have {len(numeric_features)} numeric features: {numeric_features}")
print(f"We have {len(categorical_features)} categorical features: {categorical_features}")

We have 3 numeric features: ['math score', 'reading score', 'writing score']
We have 5 categorical features: ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


##### __3.8 Finding statistics for 'score' columns__

In [40]:
df["total score"] = df['math score'] + df['writing score'] + df['reading score']
df["average score"] = df["total score"]/3
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score  average score  
0          218      72.666667  
1          247      82.333333  
2          278      92.666667  
3          148      49.333333  
4          229      76.333333

In [44]:
# Find the number of students who scored 100 for each test
reading_full = df[df['reading score'] == 100]['average score'].count()
math_full = df[df['math score']== 100].count(True)

print(reading_full)
print(math_full)

17
149    10
451    10
458    10
623    10
625    10
916    10
962    10
dtype: int64
